pkg/devicetwin/dtmanager/membership_test.go

```diff
@@ -0,0 +1,352 @@
+	ormerMock.EXPECT().Insert(gomock.Any()).Return(int64(1), nil).Times(1)
+	ormerMock.EXPECT().Commit().Return(nil).Times(1)
+	ormerMock.EXPECT().QueryTable(gomock.Any()).Return(querySeterMock).Times(1)
+	dtc := &dtcontext.DTContext{
+		DeviceList:  &sync.Map{},
+		DeviceMutex: &sync.Map{},
+		Mutex:       &sync.Mutex{},
+		GroupID:     "1",
+	}
+
+	payload := dttype.MembershipUpdate{RemoveDevices: []dttype.Device{{ID: "DeviceA", Name: "Router",
+		State: "unknown"}}, BaseMessage: dttype.BaseMessage{EventID: "eventid"}}
+	content, _ := json.Marshal(payload)
+	var m = &model.Message{
+		Content: content,
+	}
+	_, err := dealMembershipUpdated(dtc, "t", m)
+	assert.NoError(t, err)
+}
+
+func TestDealMerbershipGetEmptyMsg(t *testing.T) {
+	dtc := &dtcontext.DTContext{
+		DeviceList: &sync.Map{},
+		GroupID:    "1",
+	}
+	value, err := dealMerbershipGet(dtc, "t", "invalid")
+	assert.Error(t, err)
+	assert.Equal(t, errors.New("msg not Message type"), err)
+	assert.Equal(t, nil, value)
+}
+
+func TestDealMerbershipGetInvalidMsg(t *testing.T) {
+
+	dtc := &dtcontext.DTContext{
+		DeviceList: &sync.Map{},
+		GroupID:    "1",
+	}
+
+	var m = &model.Message{
+		Content: "hello",
+	}
+
+	value, err := dealMerbershipGet(dtc, "t", m)
+	assert.Error(t, err)
+	assert.Equal(t, errors.New("assertion failed"), err)
+	assert.Equal(t, nil, value)
+}
+
+func TestDealMerbershipGetValid(t *testing.T) {
+	dtc := &dtcontext.DTContext{
+		DeviceList:  &sync.Map{},
+		DeviceMutex: &sync.Map{},
+		Mutex:       &sync.Mutex{},
+		GroupID:     "1",
+	}
+
+	payload := dttype.MembershipUpdate{AddDevices: []dttype.Device{{ID: "DeviceA", Name: "Router",
+		State: "unknown"}}, BaseMessage: dttype.BaseMessage{EventID: "eventid"}}
+	content, _ := json.Marshal(payload)
+	var m = &model.Message{
+		Content: content,
+	}
+	_, err := dealMerbershipGet(dtc, "t", m)
+	assert.NoError(t, err)
+}
+
+func TestDealGetMembershipValid(t *testing.T) {
+	dtc := &dtcontext.DTContext{
+		DeviceList:  &sync.Map{},
+		DeviceMutex: &sync.Map{},
+		Mutex:       &sync.Mutex{},
+		GroupID:     "1",
+	}
+
+	payload := dttype.MembershipUpdate{AddDevices: []dttype.Device{{ID: "DeviceA", Name: "Router",
+		State: "unknown"}}, BaseMessage: dttype.BaseMessage{EventID: "eventid"}}
+	content, _ := json.Marshal(payload)
+
+	err := DealGetMembership(dtc, content)
+	assert.NoError(t, err)
+}
+
+func TestDealGetMembershipInValid(t *testing.T) {
+	dtc := &dtcontext.DTContext{
+		DeviceList:  &sync.Map{},
+		DeviceMutex: &sync.Map{},
+		Mutex:       &sync.Mutex{},
+		GroupID:     "1",
+	}
+
+	err := DealGetMembership(dtc, []byte("invalid"))
+	assert.NoError(t, err)
+}
+
+func TestAdded(t *testing.T) {
+	dtc := &dtcontext.DTContext{
+		DeviceList:  &sync.Map{},
+		DeviceMutex: &sync.Map{},
+		Mutex:       &sync.Mutex{},
+		GroupID:     "1",
+	}
+
+	var d = []dttype.Device{{
+		ID:    "DeviceA",
+		Name:  "Router",
+		State: "unknown",
+	}}
+	var b = dttype.BaseMessage{
+		EventID: "eventid",
+	}
+	Added(dtc, d, b, true)
```

Comment:
The function sends a message in line 279. Can we receive that message so that we can assert something here?
Similar applicable in case of removed function

Reply:
The function Added has no return type to assert.
This test case is added in order to cover the true(delta) case.

Reply:
Oh i see. Then maybe we can remove this function as without any assert we cannot be sure of testing a functionality. This function is just improving coverage correct ?
if yes, we need not have functions just to improve coverage ?
@m1093782566 agreed ?
Similar case in TestRemoved i guess.

Reply:
I have commented the code by adding comment instead of removing.

Reply:
Sounds good :)